Linting, type annotations

lk16 opened the pull request:
Hello again!

In this PR I did the following things:
* Format all code with black (respecting the 120 char lines)
* Add type annotations 
* Add pre-commit hooks for the above and a few other things (isort, ...)

Pre-commit hooks can be set up with `pre-commit install` (after `pip install pre-commit`)
After that the hooks will run automatically on each commit.
If a check fails the commit doesn't go through, so it effectively enforces that the checks pass on new commits.
Of course there is a bypass `git commit <other flags> --no-verify` which shouldn't be used IMO.

I noticed some issues:
* Some classes have members set to None in the initializer, so that field remains `Optional[...]`. I had to put some asserts in to convince mypy. These cases should probably be refactored.
* I noticed type checking did not work as strict with `immutatles.Map` as it would with `Dict` or other more common types, so please check if I got those types right.
* In a few cases I couldn't convince mypy the tying was right and I had to resort to suppressing its complaints.

This PR is quite big unfortunately.
If this gets merged I recommend you squash my commits together.

I have some ideas for other things to fix/improve, so I may add more PR's later.

Looking forward to hearing your opinion on my changes.

lk16

sashimi-houdini commented on skepticoin/signing.py:
```diff
@@ -97,28 +102,28 @@ def is_not_signature(self):
+    def __init__(self, height: int, signature: bytes):
         if not (0 <= height <= 0xFFFFFFFF):
-            raise ValueError("CoinbaseData height out of range." % height)
+            raise ValueError("CoinbaseData height %d out of range." % height)
```
It seems this point is not reached regularly :thinking:

sashimi-houdini commented on skepticoin/pow.py:
```diff
@@ -38,7 +40,12 @@ def select_block_slice(hash, serialized_block, length):
+def select_slice_from_chain(
+    input_hash: bytes,
+    current_height: int,
+    get_block_by_height: Callable[..., Serializable],
```
shouldn't this be a callable of int to block? (going by the name here)

sashimi-houdini commented on skepticoin/pow.py:
```diff
@@ -47,8 +54,12 @@ def select_slice_from_chain(input_hash, current_height, get_block_by_height, len
-):
+    starting_hash: bytes,
+    current_height: int,
+    get_block_by_height: Callable[..., Serializable],
```
same as above

sashimi-houdini commented on skepticoin/datatypes.py:
```diff
@@ -1,85 +1,95 @@
 
         if not (0 <= index <= 0xFFFFFFFF):
-            raise ValueError("OutputReference index out of range." % index)
+            raise ValueError("OutputReference index %d is out of range." % index)
```
Another good catch (either by you or by your tooling)

lk16 replied:
Either black or mypy caught this one.

sashimi-houdini commented on skepticoin/merkletree.py:
```diff
@@ -32,54 +38,59 @@ def hash(self):
     if index_of_interest >= merkle_node.children[1].index:
         other, recurse_into = merkle_node.children
-        reconstruct = lambda ot, rec: (ot, rec)  # noqa
+        reconstruct = lambda ot, rec: [ot, rec]
```
the `noqa` directive was presumably to get flake8 to shut up about inline lambda's; I'm not sure how you managed to do that if you've removed that (if at all)

sashimi-houdini commented on skepticoin/explorer.py:
```diff
@@ -24,7 +24,7 @@
 
 
-def show_coin(sashimi):
+def show_coin(sashimi: float) -> str:
```
sashimis are integers

sashimi-houdini commented on skepticoin/scripts/send.py:
```diff
@@ -74,6 +75,10 @@ def main():
             # it's late and I'm too lazy for the efficient & correct implementation.
             coinstate = thread.local_peer.chain_manager.coinstate
+
+            # TODO mypy complains coinstate could be None here
```
I'm having trouble parsing this, which of the below 2 is it?

mypy complains, that's why you added the assert and now it no longer complains

OR

mypy complained, I added this assert to get it to shut up

lk16 replied:
Not sure if I went into detail about this, so I'll do it here.

What happens is that a field (`coinstate` in this case) can be `None`, because it's declared as such in the `__init__()` function.
Therefore the field will remain `Optional[...]` for the rest of the program.
The solution would be to make the coinstate value an argument of the `__init__()` function to rule out this problematic case. I just don't want to do that in this PR, because it's out of scope.

IMO this just highlights the fact that we need to address this. It may look ugly, but in practice this assert statement has the same value as `assert True`: it will not fail and is therefore effectively redundant.

But if we don't do this then mypy will complain. And I really don't want to put `type: ignore` everywhere where `coinstate` is used, as that defeats the purpose of using type-checking.

sashimi-houdini commented on skepticoin/networking/utils.py:
```diff
@@ -1,29 +0,0 @@
-import json
```
no moving around of code while doing other things please

lk16 replied:
I had to move these 3 functions into `peers.py` to prevent an import-cycle.
This cycle came into existence when adding the type annotations.

This should be the only moved code in this PR.

sashimi-houdini commented on skepticoin/networking/peer.py:
```diff
@@ -8,82 +10,62 @@
 
 LISTENING_SOCKET = "LISTENING_SOCKET"
-IRRELEVANT = "IRRELEVANT"
+IRRELEVANT = -1
```
I understand why you do this, but please don't, as I don't feel like evaluating whether it will have real consequences and it's not worth it. Something I can think of off the top of my head: this string actually ends up in `peers.json`

sashimi-houdini commented on skepticoin/networking/peer.py:
```diff
@@ -226,7 +208,10 @@ def step(self, current_time):
-    def should_actively_fetch_blocks(self, current_time):
+    def should_actively_fetch_blocks(self, current_time: int) -> bool:
+
+        assert self.coinstate
```
I don't like this one; I'm assuming it's needed to get mypy to shut up about something, but I'd think it should be possible to do something more idiomatic instead.

lk16 replied:
Resolving this one, see other thread about this topic.

sashimi-houdini commented on skepticoin/networking/peer.py:
```diff
@@ -257,6 +242,9 @@ def add_transaction_to_pool(self, transaction):
             try:
                 validate_non_coinbase_transaction_by_itself(transaction)
+
+                assert self.coinstate
```
as elsewhere

lk16 replied:
Resolving this one, see other thread about this topic.

sashimi-houdini commented:
Ok, this is really great work... but as it stands it cannot go in. This is only partially your fault :-)

* Black is a definite no-no to me for reasons of personal preference. This is unfortunate, as it will make pulling the useful bits out of this PR much harder... but it's still non-negioble.

* In terms of mypy I'm personally on the fence, but looking through this PR it is undeniable that the code becomes much easier to reason about. So I'd say it's a definite win. But:

    * We (I) need instructions (somewhere) on how to run mypy locally. These probably need to be tied in to the other (as of yet non-existing) instructions on how to run the tests etc, but since I'm the victim this time it's a good time (for me) to make a point of it.
    * Again, personal preference: pre-commit hooks, I'd rather have as suggestions than as something that's hard requirement (this is my potentially incorrect understanding of what you've set up).
    * On the other hand, if mypy is something we mandate, it needs to go into the github actions also.

The last 2 points summarized are: validate server-side.

For the part that's not your fault: some (quite important) parts of the code are insufficiently covered by tests to take on a refactoring this big. This is something I've postponed for too long, but something that really needs to be done first. If it's something you're interested in, we can talk about how to divide the work.

And maybe a final quibble: if (when) we take on the mypy thing, I really want to have it as purely as humanly possible be about type annotations. i.e. no moving code around, no renaming etc. which could all be points of breakage that need to be checked separately.

Re your "I noticed some issues" -> yes.

sashimi-houdini commented:
> Sure, I can remove black, not sure how painful reverting the black-linting will be though. fearful
> 

I imagine (but am not sure) that cherry-picking the non-black commits semi-manually may in fact be less painful. Not sure it will be though, leaving it up to you how to solve this one.


> 
> I created issue #22: add instructions for contributing and running tests.

I've added the bits that I could think of; it would then be up to you to add the mypy instructions to that document.


> Having the `pre-commit-config.yaml` by itself does nothing until `pip install pre-commit` and `pre-commit install` are run. However I will adjust the values such that they match flake8 for the CI that is setup already. I didn't see the CI setup at the time I submitted this PR, otherwise I would have done it earlier.
> 
> It's just a tool that, when setup correctly, can prevent unnecessary CI fails, so it brings convenience.
> And if devs prefer to not use it, that's fine, CI will catch their issues anyway.

Excellent, this can then become a recommendation (rather than requirement) in the CONTRIBUTION notes.

> 
> > On the other hand, if mypy is something we mandate, it needs to go into the github actions also.
> 
> Of course.
> 

Feel free to add it.

> > For the part that's not your fault: some (quite important) parts of the code are insufficiently covered by tests to take on a refactoring this big. This is something I've postponed for too long, but something that really needs to be done first. If it's something you're interested in, we can talk about how to divide the work.
> 
> I think that refactoring _after_ adding checks (mypy + annotations) makes more sense than refactoring _before_ adding checks. Especially since you say tests are currently missing. Now I do understand that I made some changes that should not be in this huge PR.

Yeah so my worry was more about anything that's not _just_ adding type annotations.

> If I revert the changes black introduced, that would reduce this PR to:
> 
>     * adding type annotations
> 
>     * adding the `pre-commit-config.yaml` (Only flake8 and mypy)
> 
>     * adding mypy to the github CI config
> 
> 
> All other changes can be reverted. That would lead to a diff that only adds annotations in the `*.py` files and does not change any behaviour.
> 
> This will be straight-forward to review and refactoring from there will be way less painful.

Sounds great!

> > If it's something you're interested in, we can talk about how to divide the work.
> 
> If my text wall did not convince you: Sure, I am up for it, but both me and my editor will have a hard time understanding how things relate to each other without any type annotations.

I was already convinced of type annotations as presented here. :smile:

lk16 commented:
I reverted black's changes and fixed some other things as pointed out before.
I need to still add mypy to gitlab CI and rebase over the latest master.

lk16 commented:
@sashimi-houdini 

This is by far not the most interesting PR to work on for this project, but it is my oldest still-open PR.
Therefore I wanted to wrap it up before continuing the next one.
I believe this is ready to be merged now.

I ran into a few issues while getting the typing right:
* I broke a test but didn't notice until I type annotated everything. I had to revert the faulty commit, so it's fixed
* I noticed mypy didn't run the same as on CI. That's fixed now.
* Pre-commit also runs the tests now.

This PR is up-to-date with your master branch at the time of writing.

sashimi-houdini commented:
Ignore the mess above, I pushed to the wrong branch and Github won't allow me to do full cleanup.
Instead, read the below:

The squashed version of this commit made it to master in 37758d0994d5fbf83ee238faaf593a4a916043a4

I then made 3 minor changes myself (straight on master)

* dadd7c7705c3c963d735519b12c1bd74a9f9368e
* 24a72f31bc8eb1740201b60f5be028dfe98b4eb5
* d217971f1e08f7453c7d71cb74036b8b6b99aac5
